File: src/main/java/ru/citeck/idea/index/CiteckIndex.java
package ru.citeck.idea.index;

import com.intellij.util.indexing.*;
import com.intellij.util.io.DataExternalizer;
import com.intellij.util.io.KeyDescriptor;
import org.jetbrains.annotations.NotNull;

import ru.citeck.idea.files.FileType;
import ru.citeck.idea.files.FileTypeService;

import java.util.List;
import java.util.Map;

public class CiteckIndex extends FileBasedIndexExtension<IndexKey, List<IndexValue>> {

    public static final ID<IndexKey, List<IndexValue>> NAME = ID.create("ru.citeck.indexes.EcosIndexExtension");

    private final DataExternalizer<List<IndexValue>> dataExternalizer = new IndexValue.Externalizer();
    private final KeyDescriptor<IndexKey> keyDescriptor = new IndexKey.Descriptor();

    @Override
    public @NotNull ID<IndexKey, List<IndexValue>> getName() {
        return NAME;
    }

    @Override
    public @NotNull DataIndexer<IndexKey, List<IndexValue>, FileContent> getIndexer() {
        return this::map;
    }

    private @NotNull Map<IndexKey, List<IndexValue>> map(@NotNull FileContent inputData) {

        Indexes indexes = new Indexes();
        if (inputData.getFile().getFileType().isBinary()) {
            return indexes;
        }

        EcosFileIndexer.EP_NAME
            .getExtensionsIfPointIsRegistered()
            .stream()
            .filter(indexer -> {
                FileType fileType = FileTypeService.getInstance()
                    .getFileType(inputData.getFile(), inputData.getProject());
                return indexer.accept(fileType);
            })
            .forEach(ecosDataIndexer ->
                ecosDataIndexer.map(inputData, indexes.withFileContent(inputData))
            );

        return indexes;
    }

    @Override
    public @NotNull KeyDescriptor<IndexKey> getKeyDescriptor() {
        return keyDescriptor;
    }

    @Override
    public @NotNull DataExternalizer<List<IndexValue>> getValueExternalizer() {
        return dataExternalizer;
    }

    @Override
    public int getVersion() {
        return 3;
    }

    @Override
    public FileBasedIndex.InputFilter getInputFilter() {
        return file -> true;
    }

    @Override
    public boolean dependsOnFileContent() {
        return true;
    }

}

File: src/main/kotlin/ru/citeck/idea/search/index/CiteckIndex.kt
package ru.citeck.idea.search.index

import com.intellij.util.indexing.*
import com.intellij.util.io.DataExternalizer
import com.intellij.util.io.KeyDescriptor
import ru.citeck.idea.artifacts.ArtifactsService
import ru.citeck.idea.search.index.indexers.CiteckFileIndexer

class CiteckIndex : FileBasedIndexExtension<IndexKey, List<IndexValue>>() {

    private val dataExternalizer: DataExternalizer<List<IndexValue>> = IndexValue.Externalizer()
    private val keyDescriptor: KeyDescriptor<IndexKey> = IndexKey.Descriptor()

    override fun getName(): ID<IndexKey, List<IndexValue>> {
        return CiteckIndexConstants.INDEX_NAME
    }

    override fun getIndexer(): DataIndexer<IndexKey, List<IndexValue>, FileContent> {
        return DataIndexer { inputData: FileContent -> this.map(inputData) }
    }

    private fun map(inputData: FileContent): Map<IndexKey, List<IndexValue>> {
        val indexes = Indexes()
        if (inputData.file.fileType.isBinary) {
            return indexes
        }

        val artifactsService = ArtifactsService.getInstance()
        val artifactInfo = artifactsService.getArtifactInfo(inputData.file, inputData.project) ?: return indexes
        if (!artifactInfo.getController().isIndexable(inputData.psiFile)) {
            return indexes
        }

        CiteckFileIndexer.EP_NAME
            .extensionsIfPointIsRegistered
            .stream()
            .filter { it.accept(artifactInfo) }
            .forEach { it.map(inputData, artifactInfo, indexes.withFileContent(inputData)) }

        return indexes
    }

    override fun getKeyDescriptor(): KeyDescriptor<IndexKey> {
        return keyDescriptor
    }

    override fun getValueExternalizer(): DataExternalizer<List<IndexValue>> {
        return dataExternalizer
    }

    override fun getVersion(): Int {
        return 4
    }

    override fun getInputFilter(): FileBasedIndex.InputFilter {
        return FileBasedIndex.InputFilter { true }
    }

    override fun dependsOnFileContent(): Boolean {
        return true
    }
}
